multiview/cotraining/ctclassifier.py

```diff
@@ -0,0 +1,369 @@
+        performed effectively.
+
+    estimator2 : classifier object, default (sklearn GaussianNB)
+        The classifier object which will be trained on view 2 of the data.
+        Does not need to be of the same type as estimator1, but should support
+        predict_proba().
+
+    Attributes
+    ----------
+    estimator1 : classifier object
+        The classifier used on view 1.
+
+    estimator2 : classifier object
+        The classifier used on view 2.
+
+    class_name: string
+        The name of the class.
+
+    n_views_ : int
+        The number of views supported by the multi-view classifier
+
+    classes_ : array-like of shape (n_classes,)
+        Unique class labels.
+
+    n_classes_ : int
+        Number of unique classes.
+
+    p_ : int
+        Number of positive examples (second label in classes_) to pull from
+        unlabeled_pool and give "label" at each training round. When the
+        labels are 0 or 1, positive is defined as 1, and in general, positive
+        is the larger label.
+
+    n_ : int
+        Number of negative examples (second label in classes_) to pull from
+        unlabeled_pool and give "label" at each training round. when the
+        labels are 0 or 1, negative is defined as 0, and in general, negative
+        is the smaller label.
+
+    unlabeled_pool_size_ : int
+        Size of pool of unlabeled_pool examples to classify at each iteration.
+
+    num_iter_ : int
+        Maximum number of training iterations to run.
+
+    random_state : real number
+        The starting random seed for fit() and class operations.
+
+    References
+    ----------
+    [1] Blum, A., & Mitchell, T. (1998, July). Combining labeled and
+    unlabeled_pool data with co-training. In Proceedings of the eleventh
+    annual conference on Computational learning theory (pp. 92-100). ACM.
+
+    """
+
+    def __init__(
+                 self,
+                 estimator1=None,
+                 estimator2=None,
+                 random_state=0
+                 ):
+
+        # initialize a BaseCTEstimator object
+        super().__init__(estimator1, estimator2, random_state)
+
+        # if not given, set classifiers as gaussian naive bayes estimators
+        if self.estimator1 is None:
+            self.estimator1 = GaussianNB()
+        if self.estimator2 is None:
+            self.estimator2 = GaussianNB()
+
+        # verify that estimator1 and estimator2 have predict_proba
+        if (not hasattr(self.estimator1, 'predict_proba')
+                or not hasattr(self.estimator2, 'predict_proba')):
+            raise AttributeError("Co-training classifier must be initialized "
+                                 "with classifiers supporting "
+                                 "predict_proba().")
+
+        self.n_views_ = 2  # only 2 view learning supported currently
+
+        self.class_name = "CTClassifier"
+
+    def fit(
+            self,
+            Xs,
+            y,
+            p=None,
```

Comment:
see my comment in slack, passing parameters here is not sklearn API compliant

Reply:
I can change unlabeled_pool_size and num_iter to be passed into _init_(). However, p and n are data dependent (they depend on the number of samples of each class exist in the data Xs), and if the user doesn't specify one or both of them, then they are set based on the data. Should I still pass them into _init_ and then just change them based on the data once the user calls fit? Or should I leave them here? I thought sklearn could take parameters to fit if they are data dependent (e.x. the fit method for RandomForestClassifier takes sample weights as an optional parameter).

Reply:
to me, this is more akin to RandomForestClassifier `max_features` for example. From their docs, 
 > If “auto”, then max_features=sqrt(n_features)

So the parameter does depend on the data in the sense that it depends on the dimension of the data, but I don't see that as something like `sample_weights` where you really have a unique piece of information for each sample. So if it were me, and if I'm understanding you correctly, I'd still have them be parameters in `__init__()`, but if you think it makes a lot more sense for the user the other way then go for it.

Reply:
That makes sense. I changed it so they are all passed into __init__() only. This change, along with new unit tests and the updated simulated testing notebooks are all packaged into a single PR now: https://github.com/NeuroDataDesign/multiview/pull/44